### PR TITLE
feat(weight-free): weight-free ONNX export for causal LMs (dynamo path)

### PR DESCRIPTION
**Pull request by amarquic**
## Summary

Adds weight-free ONNX export for causal LM models on top of the dynamo export infrastructure in #1178.

Weight-free export builds the model on meta device (no weights in RAM), exports ONNX graph structure only, and lets the QAIC compiler load weights directly from the original safetensors checkpoint at compile time via `weight_spec.json`.

### New files
- `QEfficient/exporter/weight_free.py` — core weight-free export: promotes ONNX initializers to external data, builds `weight_spec.json` (5 checkpoint key lookup strategies including Mixtral `mlp`→`block_sparse_moe` rename)
- `QEfficient/exporter/weight_spec.py` — weight spec format and I/O
- `QEfficient/exporter/checkpoint_transforms.py` — checkpoint prep pipeline: MoE expert stacking, dtype conversion, `.bin`→safetensors auto-convert
- `examples/text_generation/weight_free_export_verify.py` — single-model smoke test
- `examples/text_generation/continuous_batching_weightfree.py` — CB weight-free smoke test
- `examples/performance/compute_context_length/weight_free_ccl_verify.py` — CCL verification

### Key changes
- `QEfficient/base/modeling_qeff.py` — `use_weight_free_export` flag; meta-device guard for `_model_offloaded_check()`
- `QEfficient/transformers/models/modeling_auto.py` — `use_weight_free_export` wired through `export()` and `compile()`; CCL `comp_ctx_lengths` Dim fix
- `QEfficient/base/onnx_transforms.py` — disable `RewriteUnsupportedOpsTransform` and `SplitMultiInputMaxMinTransform` for dynamo path
- `QEfficient/utils/export_utils.py` — dynamo guard on `apply_torch_patches()`; separate dynamo/TorchScript subfunction setup paths
- `QEfficient/utils/torch_patches.py` — `temporarily_enable/disable_nested_compile_regions()`
- `QEfficient/customop/` — backward stubs for `CtxScatter/Gather`; `GemmaRMSNorm` weight offset at runtime
- Model fixes: Mixtral meta guard + attention_scaling, GraniteMoe optimizeSoftmax (router + attention), Falcon meta guard, GPT-OSS meta guard + device fixes, Qwen3-MoE, GLM4-MoE

## Test plan

- [ ] Weight-free export + compile + inference verified on: GPT2, Llama-3.2-1B, Mistral-7B, Qwen2-1.5B, Qwen3-30B-A3B, Phi-3-mini, Gemma-2-2b, OLMo2-1B, Falcon-7B, GPT-J-6B, Starcoder2-3B, GraniteForCausalLM, GraniteMoE, GLM4-MoE, GPT-OSS-20B, Mixtral-8x7B, Qwen3-MoE
- [ ] CCL verified on Qwen3-30B-A3B
- [ ] Continuous batching weight-free verified on Qwen2-1.5B, Llama-3.2-1B

### CODE REVIEW

File: QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py [hunk 1]
```diff
@@ -204,11 +208,22 @@ def forward(self, hidden_states):
...
+            E = self.gate_up_proj.shape[0]
+            H = self.gate_up_proj.shape[2]
+            self.gate_proj   = nn.Parameter(torch.empty(E, H, self.expert_dim, device="meta"))
+            self.up_proj     = nn.Parameter(torch.empty(E, H, self.expert_dim, device="meta"))
+            self.down_proj_t = nn.Parameter(torch.empty(E, self.expert_dim, H, device="meta"))
+            return
+
+        # Normal export: compute derived params — values embedded in ONNX.
+        self.gate_proj   = nn.Parameter(self.gate_up_proj[:, : self.expert_dim, :].detach().clone().transpose(1, 2))
+        self.up_proj     = nn.Parameter(self.gate_up_proj[:, self.expert_dim :, :].detach().clone().transpose(1, 2))
```
Comment (vbaddi) [inline QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py]: nit: why did we do clone for normal export, when original was just `.detach()` and without `.clone()`?
same question for all moe related changes.

File: examples/performance/compute_context_length/weight_free_ccl_verify.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python3
```
Comment (vbaddi) [inline examples/performance/compute_context_length/weight_free_ccl_verify.py:1]: nit: Can you add a readme (.md under /docs) on how this weightfree approach works? details on how to enable it? How download/conversion happens? and conversion being onetime thing, How we can reuse the .safetensors across machines and runs?

File: QEfficient/exporter/weight_free/spec.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+#
```
Comment (vbaddi) [inline QEfficient/exporter/weight_free/spec.py:1]: Let's create a 'weightfree/' directory and push these weightfree related things under it?

File: QEfficient/customop/ctx_scatter_gather_cb.py [hunk 4]
```diff
@@ -92,6 +97,11 @@ def forward(data: torch.Tensor, batch_index: torch.Tensor, position_ids: torch.T
     def setup_context(ctx, inputs, outputs):
         pass
 
+    @staticmethod
+    def backward(ctx, grad_output):
```
Comment (vbaddi) [inline QEfficient/customop/ctx_scatter_gather_cb.py]: nit: if we call `model.eval()`, is backward expected? Are these changes required?

File: scripts/preprocess_moe_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
+# -----------------------------------------------------------------------------
```
Comment (vbaddi) [inline scripts/preprocess_moe_checkpoint.py:1]: unit test infra is missing, pls work @smedhe and @quic-amitraj to get this added to avoid manual runs.

**Review by vbaddi — CHANGES_REQUESTED** [review]

**Comment by vbaddi** [comment]
lint and format are missing. Pls follow the workflows to enable and repush once its fixed.

File: QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py (hunk 1 above)
Comment (Cs23m011) [reply QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py]: We are seeing the aliasing issue in dynamo without clone () method. It will not export in dynamo without clone as it will get nodes with same name. And in weight free export we are not using this the else block anyway.

File: examples/performance/compute_context_length/weight_free_ccl_verify.py (hunk 2 above)
Comment (amarquic) [reply examples/performance/compute_context_length/weight_free_ccl_verify.py:1]: Okay I will add that.

File: QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py (hunk 1 above)
Comment (vbaddi) [reply QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py]: Re: _And in weight free export we are not using this the else block anyway._ If not then we should keep it same as before right?

Comment (amarquic) [reply QEfficient/transformers/models/qwen3_moe/modeling_qwen3_moe.py]: But if user uses Weight_free_export=false then it will go via dynamo path and changing that will fail dynamo export because of aliasing issue.

File: examples/performance/compute_context_length/weight_free_ccl_verify.py (hunk 2 above)
Comment (amarquic) [reply examples/performance/compute_context_length/weight_free_ccl_verify.py:1]: Done.

File: QEfficient/exporter/weight_free/spec.py (hunk 3 above)
Comment (amarquic) [reply QEfficient/exporter/weight_free/spec.py:1]: Done. Please review.

File: QEfficient/customop/ctx_scatter_gather_cb.py (hunk 4 above)
Comment (amarquic) [reply QEfficient/customop/ctx_scatter_gather_cb.py]: Removed. Please review.

**Comment by vbaddi** [comment]
duplicate #1220
